### ISSUE TITLE
Variance bug with `__add__` and `list` generic

### ISSUE THREAD

**Issue by not-my-profile** [microsoft/pyright#3110]
```python
class Node:
    def __add__(self, other: list['Node']):
        pass

class Element(Node):
    pass

Element() + [Element()]
```

Pyright reports an incorrect error for the last line:

> Operator "+" not supported for types "Element" and "list[Element]"

Mypy and pytype handle the example correctly by not reporting any errors.

A similiar example that pyright handles incorrectly is:

```python
class Node:
    def foo(self, other: list['Node']) -> list['Node']:
        return [self] + other
```

here pyright reports:
```
Expression of type "list[Self@Node]" cannot be assigned to return type "list[Node]"
  TypeVar "_T@list" is invariant
    Type "Self@Node" cannot be assigned to type "Node"
```

whereas mypy and pytype don't report any errors.

**Comment by erictraut** [comment]
This behavior is correct. You have annotated the `Node` method to accept a parameter of type `list[Node]`, but the type parameter for `list` is invariant, so only `list[Node]` will be accepted here. In your usage, you're attempting to add a `list[Element]`, which represents a type violation.

I recommend changing the signature to:
```python
    def __add__(self, other: Sequence["Node"]):
```
